src/blob/core.rs

```diff
@@ -350,51 +358,61 @@ where
         key: &K,
         meta: Option<&Meta>,
         check_filters: bool,
-    ) -> Result<Option<Entry>> {
+    ) -> Result<ReadResult<Entry>> {
         debug!("blob get any entry {:?}, {:?}", key, meta);
         if check_filters && self.check_filter(key).await == FilterResult::NotContains {
             debug!("Key was filtered out by filters");
-            Ok(None)
+            Ok(ReadResult::NotFound)
         } else if let Some(meta) = meta {
             debug!("blob get any entry meta: {:?}", meta);
             self.get_entry_with_meta(key, meta).await
         } else {
             debug!("blob get any entry bloom true no meta");
-            if let Some(header) = self.index.get_any(key).await.with_context(|| {
-                format!("index get any failed for blob: {:?}", self.name.to_path())
-            })? {
-                let entry = Entry::new(header, self.file.clone());
-                debug!("blob, get any entry, bloom true no meta, entry found");
-                Ok(Some(entry))
-            } else {
-                Ok(None)
-            }
+            Ok(self
+                .index
+                .get_any(key)
+                .await
+                .with_context(|| {
+                    format!("index get any failed for blob: {:?}", self.name.to_path())
+                })?
+                .map(|header| {
+                    let entry = Entry::new(header, self.file.clone());
+                    debug!("blob, get any entry, bloom true no meta, entry found");
+                    entry
+                }))
         }
     }
 
-    async fn get_entry_with_meta(&self, key: &K, meta: &Meta) -> Result<Option<Entry>> {
+    async fn get_entry_with_meta(&self, key: &K, meta: &Meta) -> Result<ReadResult<Entry>> {
         let headers = self.index.get_all(key).await?;
         if let Some(headers) = headers {
             let entries = Self::headers_to_entries(headers, &self.file);
             self.filter_entries(entries, meta).await
         } else {
-            Ok(None)
+            Ok(ReadResult::NotFound)
         }
     }
 
-    async fn filter_entries(&self, entries: Vec<Entry>, meta: &Meta) -> Result<Option<Entry>> {
+    async fn filter_entries(&self, entries: Vec<Entry>, meta: &Meta) -> Result<ReadResult<Entry>> {
         for mut entry in entries {
             if Some(meta) == entry.load_meta().await? {
-                return Ok(Some(entry));
+                return Ok(if entry.header().is_deleted() {
+                    ReadResult::Deleted(entry.header().created())
+                } else {
+                    ReadResult::Found(entry)
+                });
             }
         }
-        Ok(None)
+        Ok(ReadResult::NotFound)
     }
 
-    pub(crate) async fn contains(&self, key: &K, meta: Option<&Meta>) -> Result<bool> {
+    pub(crate) async fn contains(&self, key: &K, meta: Option<&Meta>) -> Result<ReadResult<u64>> {
```

Comment:
This is not an obvious change. You must create a wrapper struct (or typedef) to emphasize that this `u64` is the creation timestamp